Add Kafka-Humio Gateway [Depends on PR#10]

jeffmahoney opened the pull request:
This PR adds a Kafka plugin that can be used to export events to a Kafka instance. The instance is expected to be running in the same container ecosystem as the velociraptor server. We do this so that we can ensure that events are queued properly if the Humio server can't be reached. Once the events are queued, a consumer formats and forwards them to Humio.

jeffmahoney committed:
Add a Kafka export plugin

This plugin works similarly to the Elastic plugin but allows the
Velocraptor server to export events and artifacts to an in-cluster
Kafka server for guaranteed delivery.  The events can then be consumed
by an agent and then ingested into a SIEM or other system.

jeffmahoney committed:
contrib: Add kafka-humio gateway

This separate helper will consume events posted to Kafka, transform
them, and post them to Humio.

Firstyear reviewed (commented):
This all looks good to me, but as I'm not a go-expert, I don't feel confident to give the approve.

jeffmahoney commented:
I think the windows failure is a testcase issue. There's a timer that trips after 2 minutes and dumps state. It's reset for each test. These changes don't touch anything that would be involved there.

lorddoskias commented on contrib/kafka-humio-gateway/go.mod:
```diff
@@ -0,0 +1,8 @@
+module main
+
+go 1.16
```
I wonder if we can bump this to 1.17 given velociraptor itself depends on go 1.17. That's not a deal breaker but just if we want to keep the two of them in sync. Feel free to disregard.

jeffmahoney replied:
The 0.6.3 branch depends on go 1.16.

lorddoskias commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+}
+
+func init() {
+	flag.BoolVar(&verbose, "verbose", false, "Sarama logging")
```
"Sarama logging" is supposed to be the usage string, just having "Sarama logging" seems kind of ambiguous, perhaps a short explanation what does verbose enable would be better.

lorddoskias commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+	"syscall"
+	"time"
+
+	"github.com/go-yaml/yaml"
```
I assume this is working but going over to https://github.com/go-yaml/yaml it says the import part should be: 
gopkg.in/yaml.v2

I wonder if this could have some reprcussion in the future.

jeffmahoney replied:
It does, but it's also unclear who owns and maintains gopkg.in. I'd prefer to stick with GitHub as the anchor.

lorddoskias commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+			if ctx.Err() != nil {
+				return
+			}
+			consumer.ready = make(chan bool)
```
This will be re-initializing the channel on every iteration, is this intended? Looking at the <-consumer.ready further down I'd expect to only want to initialize the consumer.ready chan just once?

lorddoskias replied:
Actually consumer.ready is already initialized at the top of the function and I don't see where anything is being written to it. So perhaps you mean to do consumer.ready <- true here so unblock the main thread?

lorddoskias replied:
Ok, this is actually correct since the channel is closed in the setup() callback of the consumer which would unblock the later read from the channel, thus ensuring the main thread continues only after the consumer has been initialized. However I find this rather non-obvious and probably a comment explaining this would be useful.

jeffmahoney replied:
TBH, I pulled this from a ConsumerGroup example. I agree it could be clearer.

jeffmahoney replied:
I'm going to replace this with a bool and a WaitGroup rather than this mess. The only reason the channel is re-made is because the setup callback expects it to be live. Nothing else uses it.

lorddoskias commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+		}
+
+		payload := HumioPayload{
+			Events: []HumioEvent {
```
Here we only ever have a single event?

jeffmahoney replied:
Yes, but the API for Humio requires that the events be submitted in a JSON array. Eventually we could submit multiple events in a single API call, but that would require that all events in the same submission use the same tags. It's unclear how we want to use tags generally, but for now we're using hostname, client id, and VRR artifact. We could batch them but I'm not sure how much benefit that would bring compared with the complexity.

lorddoskias replied:
Fair enough.

jeffmahoney replied:
I'm going to revisit this. As I started enabling noisier artifacts like process execution and network activity, it became obvious we will need to implement some kind of batching here.

jeffmahoney replied:
I've revisited this in commit [b8be7d22](https://github.com/SUSE/linux-security-sensor/pull/8/commits/b8be7d22b7c76f375a8c73ddcb6b55e2ed957f63).

lorddoskias commented:
Overall it looks good I only have a couple of minor nits which I've put inline. For an initial version I think it's good to go.

jeffmahoney commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+		sarama.Logger = log.New(os.Stdout, "[sarama] ", log.LstdFlags)
+	}
+
+	body, err := os.ReadFile(configFile)
```
That needs error handling

jeffmahoney commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+
+	body, err := os.ReadFile(configFile)
+
+	err = yaml.Unmarshal(body, &consumer.config)
```
This too

jeffmahoney committed:
kafka-humio-gateway: Add missing error handling

The file open and yaml parsing calls didn't have error handling.

jeffmahoney commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -0,0 +1,288 @@
+	ready chan bool
+	httpClient http.Client
+	config TransportConfig
+	performTransformations bool
```
This isn't used anymore

jeffmahoney committed:
kafka-humio-gateway: remove unused Consumer member

The performTransformations member was a relic of the abandoned
transformation feature.

jeffmahoney committed:
kafka-humio gateway: improve argument parsing and messages

jeffmahoney committed:
kafka-humio-gateway: make setup-and-wait race more obvious

The channel method is essentially an open-coded waitgroup that must
be tended for little benefit.  Let's just replace it with an actual
waitgroup.

jeffmahoney committed:
kafka-humio-gateway: cleanups to move related things closer together

jeffmahoney reviewed (commented):
Thanks for the review, Nik. I'll resubmit with the changes.

lorddoskias commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -162,15 +166,19 @@ func main() {
+		// Allow the main thread to proceed
+		consumer.readyWg.Done()
+	}
+	consumer.setupOnce = true
```
All of this can be replaced with an https://pkg.go.dev/sync#Once

jeffmahoney replied:
Yep, that would be nicer.

jeffmahoney committed:
kafka-humio-gateway: use sync.Once instead of open coding it

jeffmahoney commented on contrib/kafka-humio-gateway/main.go:
```diff
@@ -165,19 +165,15 @@ func main() {
-		consumer.readyWg.Done()
-	}
-	consumer.setupOnce = true
+	consumer.setupOnce.Do(consumer.readyWg.Done())
```
This is wrong.

jeffmahoney committed:
kafka-humio-gateway: fix typo in previous commit

The sync.Once API needs the go equivalent of a function pointer not
the result of the function call.

jeffmahoney committed:
kafka-humio-gateway: fix handling of timestamps

Humio only expects to provide the timezone if the timestamp is provided
as a numeric value and throws an error if one is provided anyway.

jeffmahoney committed:
kafka-humio-gateway: whitespace fixes

Wrap lines at 100 columns.

jeffmahoney committed:
kafka-humio-gateway: fix detection of epoch-style timestamp

The detection of an epoch-style timestamp that would also require
specifying a timezone reused the 'ok' variable from the previous
statement, which tripped the validity check for the event.

jeffmahoney committed:
kafka-humio-gateway: rework submission to batch events

The existing implementation worked fine when it was just pushing the
occasional login event.  Now that we have network connection events
and process execution events, the event count has increased
substantially.  Humio supports the batching of events in a single POST
submission, so we do that now.  We'll batch either 500 events or will
submit every 3 seconds.  Both numbers are configurable.  As of this
writing Humio has a 32MB limit for a single POST.

lorddoskias reviewed (commented):
For such type of fixes I suggest you use 'go fmt' command as it's supposed to just do the right thing. Perhaps we can make it a pre-commit hook .

lorddoskias reviewed (approved):
On average how many messages would be in a single claim?  I was wondering if we could do the batching on claim granularity i.e not have sendEvent and the eventChannel but simply buffer all mesages from a single claim and make that be the batch. I guess we could end up in situation where we have multiple claims consisting of a single message? Anyways this is just a remark and the changes LGTM

jeffmahoney commented:
> On average how many messages would be in a single claim? I was wondering if we could do the batching on claim granularity i.e not have sendEvent and the eventChannel but simply buffer all mesages from a single claim and make that be the batch. I guess we could end up in situation where we have multiple claims consisting of a single message? Anyways this is just a remark and the changes LGTM

The Claim isn't a one-shot. It starts up a goroutine that feeds messages from Kafka from each claim until the context is canceled or the server rebalances (and then it'll just reconnect). That's also why sendEvents is a separate goroutine -- we want to push events on both a timeout and a batch limit, and we can't do that if we're waiting on the for loop in ConsumeClaim.
